### PR TITLE
JSPI - Support EM_*JS functions.

### LINKED ISSUE #17890: Missing function signature when using EM_ASYNC_JS and ASYNCIFY=2
Error message:
`Aborted(Assertion failed: Missing __sig for __asyncjs__promise42)`

Building with:
`emcc -sASSERTIONS -sENVIRONMENT=web -g3 -sASYNCIFY=2 -sEXIT_RUNTIME=1 main.cpp -o a.out.html`

main.cpp
```c
#include <stdio.h>
#include <emscripten.h>

EM_ASYNC_JS(int, promise42, (), {
  return new Promise((k,r)=>{setTimeout(()=> {return k(42)},0)});
});

int main() {
  printf("p1\n");
  promise42();
  printf("p2\n");
  return 0;
}

```

I need to look into why the signatures aren't being generated for EM_ASYNC_JS functions.

### PR DESCRIPTION
**Pull request by brendandahl**
Generate the needed type signatures to create wrapper WebAssembly Functions.

Fixes #17890

### CODE REVIEW

File: test/test_core.py [hunk 1]
```diff
@@ -9494,9 +9494,13 @@ def test_embind_lib_with_asyncify(self, args):
...
   @no_asan('asyncify stack operations confuse asan')
   @no_wasm64('TODO: asyncify for wasm64')
+  @with_asyncify_and_stack_switching
   def test_em_async_js(self):
     self.uses_es6 = True
-    self.set_setting('ASYNCIFY')
+    if not self.get_setting('ASYNCIFY'):
+      self.set_setting('ASYNCIFY')
+    else:
+      self.set_setting('EXIT_RUNTIME')
```
Comment (sbc100) [inline test/test_core.py:9503]: Is EXIT_RUNTIME always needed here?  Why?

Also, why not just unconditionally set both of these here?

File: tools/extract_metadata.py [hunk 2]
```diff
@@ -288,6 +290,12 @@ def extract_metadata(filename):
           invoke_funcs.append(i.field)
         elif i.field not in em_js_funcs:
           declares.append(i.field)
+        elif i.field in em_js_funcs:
+          em_js_funcs_type_id[i.field] = i.type
+
+    types = module.get_types()
+    for type in em_js_funcs_type_id.values():
+      em_js_func_types[type] = types[type]
```
Comment (sbc100) [inline tools/extract_metadata.py:298]: Can you wrap these three lines in `if em_js_funcs_type_id:`?

File: src/library_async.js [hunk 3]
```diff
@@ -168,7 +173,7 @@ mergeInto(LibraryManager.library, {
             original = new WebAssembly.Function(
               { parameters , results: ['externref'] },
               original,
-              { promising : 'first' });
+              { 'promising' : 'first' });
```
Comment (sbc100) [inline src/library_async.js:176]: Is this an unrelated fix?

File: src/library_async.js [hunk 4]
```diff
@@ -89,9 +89,14 @@ mergeInto(LibraryManager.library, {
             // Wrap all imports with a suspending WebAssembly function.
             // TODO: wrap only async functions.
 #if ASSERTIONS
-            assert(sig, 'Missing __sig for ' + x);
+            assert(original.type || sig, 'Missing __sig for ' + x);
```
Comment (sbc100) [inline src/library_async.js:92]: Why not just stick to using `.sig` instead of adding another `.type` attribute?

File: emscripten.py [hunk 5]
```diff
@@ -586,6 +605,9 @@ def create_em_js(metadata):
     arg_names = [arg.split()[-1].replace("*", "") for arg in args if arg]
     args = ','.join(arg_names)
     func = f'function {name}({args}) {body}'
+    if settings.ASYNCIFY == 2 and name in metadata['emJsFuncsTypeId']:
+      typeId = metadata['emJsFuncsTypeId'][name]
+      func = func + f'\n{name}.type = EM_JS_FUNC_TYPES[{typeId}];'
```
Comment (sbc100) [inline emscripten.py:610]: Why not just use the existing `.sig = <string` mechanism that we already have?   I seem more compact and we already have `sigToWasmTypes` converting to this format at runtime.

Comment (brendandahl) [reply emscripten.py:610]: Oh yeah I wanted to ask you about this. With the metadata extraction we only get the wasm type so we lose information about pointers, so the `sig` won't really match one of the hardcoded ones we have. It seems like that could lead to some confusion down the road. Fine with me though if you aren't concerned.

File: src/library_async.js (hunk 3 above)
Comment (brendandahl) [reply src/library_async.js:176]: The new test that is enabled runs with closure and closure will rename this if it isn't in quotes.

Comment (sbc100) [reply src/library_async.js:176]: I see... what about the first argument above though?  Why does that get away without quotes?  I wonder if there is something we you modify in the closure extern files to avoid the need for this?

File: test/test_core.py (hunk 1 above)
Comment (brendandahl) [reply test/test_core.py:9503]: `with_asyncify_and_stack_switching` sets ASYNCIFY=2, it was getting overridden by doing `set_setting('ASYNCIFY')`. Also, ASYNCIFY==2 requires EXIT_RUNTIME. 
I suppose it'd be more clear if I did something like
```
if self.get_setting('ASYNCIFY') == 2:
  self.set_setting('EXIT_RUNTIME')
else:
  self.set_setting('ASYNCIFY')
```

File: emscripten.py (hunk 5 above)
Comment (brendandahl) [reply emscripten.py:610]: @sbc100 what are you thoughts here?

Comment (sbc100) [reply emscripten.py:610]: I think having `sig` without the pointer information is fine in this case.  At least no worse that what we have today for EM_JS functions which is not signature at all.  

EM_JS function cannot have explicit user-specified sig attributes since that only works for functions defined in library files (I think), so you don't need to worry about conflicts with a user-specified sig.

File: test/test_core.py (hunk 1 above)
Comment (brendandahl) [reply test/test_core.py:9503]: I did a bit of digging and the `EXIT_RUNTIME` looks to be a requirement of the old stack switching code. Seems fine without it now.

File: tools/webassembly.py [hunk 6]
```diff
@@ -79,6 +79,17 @@ def helper(self, *args, **kwargs):
...
 
+def type_to_sig(type):
+  # These should match the conversion in $sigToWasmTypes.
+  return {
+    Type.I32: 'i',
+    Type.I64: 'j',
+    Type.F32: 'f',
+    Type.F64: 'd',
+    Type.VOID: 'v'
+  }[type]
```
Comment (sbc100) [inline tools/webassembly.py:90]: Maybe put this in extract_metadata.py or emscripten.py?   webassembly.py is supposed to be just core wasm stuff.

File: tools/extract_metadata.py [hunk 7]
```diff
@@ -288,6 +289,9 @@ def extract_metadata(filename):
           invoke_funcs.append(i.field)
         elif i.field not in em_js_funcs:
           declares.append(i.field)
```
Comment (sbc100) [inline tools/extract_metadata.py:291]: Maybe flip these blocks to the condition is in the positive?

**Review by sbc100 — APPROVED** [review]

**Review by sbc100 — APPROVED** [review]
Great!
